### PR TITLE
Check port conflict with pending containers

### LINKED ISSUE #2499: Scaling up a service with mapped ports encounters port conflict errors even though there are nodes available to schedule the containers
Using classic `Swarm: 1.2.5` cluster, with 1 controller & 5 worker nodes.
I’m using docker-compose 1.8.0 to up a very simple app, nginx that has port 80 mapped to a specific port on the host. It connects to an external overlay network as below:

```
version: "2"
services:
  nginxtest:
    image: nginx
    ports:
      - "9876:80"

networks:
  default:
    external:
      name: 9498-gssp-ol-web
```

The app comes up with no issue, obviously.
But when I scale it up, I tried a scale factor of 5 and 4, I get errors about port conflicts. I have 5 nodes in total with nothing else running on them. Also I see this error below about `endpoint join on network failed` for each failed container.

```
Error scaling container
Error response from daemon: Error response from daemon: container 4e3324a0adb683357ca2f21cf8d73e097bb2857fd4cd1925a85f2669f0828b46: endpoint join on GW Network failed: driver failed programming external connectivity on endpoint gateway_4e3324a0adb6 (b1de5f11c7441993d67bc9b26ddc61694ce34301c72ffa2b0809b8653da1d376): Bind for 0.0.0.0:10000 failed: port is already allocated
```

Also, when I check the application itself, there are n+1 number of containers, n being the scale factor I had used and of course one or many of them have failed due to the port conflict above.

Now, shouldn't port filter be implicitly handled, even if I map the port explicitly and when I have more nodes than the number I want to scale the containers to? I have 5 nodes in the cluster. If I try to scale it to 3 it seems to work. Subsequently, when I change the scale factor to 4 or 5, I get a bunch of errors about the `endpoint join on GW Network failed`.

```
[ec2-user@ip-10-1-1-75 test]$ docker-compose ps
Name   Command   State   Ports
------------------------------
[ec2-user@ip-10-1-1-75 test]$ docker-compose scale nginxtest=5
WARNING: The "nginxtest" service specifies a port on the host. If multiple containers for this service are created on a single host, the port will clash.
Creating and starting test_nginxtest_1 ... done
Creating and starting test_nginxtest_2 ... done
Creating and starting test_nginxtest_3 ... done
Creating and starting test_nginxtest_4 ... error
Creating and starting test_nginxtest_5 ... done

ERROR: for test_nginxtest_4  Cannot start service nginxtest: Error response from daemon: container d3e2afc07af6b9443ea2617c64605c7cdd88e108a9ad1360994334394340db0e: endpoint join on GW Network failed: driver failed programming external connectivity on endpoint gateway_d3e2afc07af6 (81b995336de0d9dc58b81b00d2365514e15172f89b2c93c3efe75edc292e7eb4): Bind for 0.0.0.0:10003 failed: port is already allocated
[ec2-user@ip-10-1-1-75 test]$ docker-compose ps
      Name               Command           State                  Ports
-------------------------------------------------------------------------------------
test_nginxtest_1   nginx -g daemon off;   Up         443/tcp, 10.1.1.70:10003->80/tcp
test_nginxtest_2   nginx -g daemon off;   Up         443/tcp, 10.1.1.75:10003->80/tcp
test_nginxtest_3   nginx -g daemon off;   Up         443/tcp, 10.1.1.73:10003->80/tcp
test_nginxtest_4   nginx -g daemon off;   Exit 128
test_nginxtest_5   nginx -g daemon off;   Up         443/tcp, 10.1.1.72:10003->80/tcp
[ec2-user@ip-10-1-1-75 test]$
```

### PR DESCRIPTION
**Pull request by dongluochen**
Swarm checks port conflict with existing containers at create time. In concurrent container creation, it should also check **pending** containers which are under construction. This corner case is amplified by Docker compose.   

Fix #2499. 

Signed-off-by: Dong Chen dongluo.chen@docker.com

### CODE REVIEW

**Comment by dongluochen** [comment]
Ping @nishanttotla, @allencloud. 

Test failure is tracked by #2575, not related to this change.

File: scheduler/filter/port.go [hunk 1]
```diff
@@ -80,17 +80,30 @@ func (p *PortFilter) portAlreadyInUse(node *node.Node, requested nat.PortBinding
...
+		//    This corner case is amplified by Docker compose 'scale' command.
+		//    See https://github.com/docker/swarm/issues/2499.
+
+		if c.ID != "" {
+			if (c.Info.HostConfig != nil && p.compare(requested, c.Info.HostConfig.PortBindings)) || (c.Info.NetworkSettings != nil && p.compare(requested, c.Info.NetworkSettings.Ports)) {
+				return true
+			}
+		} else {
+			// container is a pending container, check its configuration
+			if c.Config != nil && p.compare(requested, c.Config.HostConfig.PortBindings) {
```
Comment (nishanttotla) [inline scheduler/filter/port.go]: @dongluochen is this working for you in testing? We should be extra careful about using `c.Config`, `c.Info.HostConfig` etc, since several of these structures are replicated in different places (unfortunately).

Otherwise code LGTM.

Comment (dongluochen) [inline scheduler/filter/port.go]: @nishanttotla Thanks for raising this question. 

For pending containers, the `c.Config` has the port info.  
https://github.com/docker/swarm/blob/master/cluster/swarm/cluster.go#L41
